tests/integration/conftest.py

```diff
@@ -191,3 +201,161 @@ async def kubernetes_cluster(request: pytest.FixtureRequest, ops_test: OpsTest):
         bundle.switch(charm.app_name, path)
     async with deploy_model(request, ops_test, model, bundle) as the_model:
         yield the_model
+
+
+@pytest_asyncio.fixture(scope="module")
+async def grafana_agent(ops_test: OpsTest, kubernetes_cluster: Model):
+    """Deploy Grafana Agent."""
+    with ops_test.model_context("main") as model:
+        await model.deploy("grafana-agent", channel="stable")
+        await model.integrate("grafana-agent:cos-agent", "k8s:cos-agent")
+        await model.integrate("grafana-agent:cos-agent", "k8s-worker:cos-agent")
+        await model.integrate("k8s:cos-worker-tokens", "k8s-worker:cos-tokens")
+
+        yield
+
+        await model.remove_application("grafana-agent")
+
+
+@pytest_asyncio.fixture(scope="module")
+async def cos_model(ops_test: OpsTest, kubernetes_cluster: Model, grafana_agent):
+    """Create a COS substrate and a K8s model."""
+    log.info("Creating COS substrate...")
+    container_name = "cos-substrate"
+    network_name = "cos-network"
+    manager = LXDSubstrate(container_name, network_name)
+
+    config = manager.create_substrate()
+    kubeconfig_path = ops_test.tmp_path / "kubeconfig"
+    kubeconfig_path.write_text(config)
+    config = type.__call__(Configuration)
+    k8s_config.load_config(client_configuration=config, config_file=str(kubeconfig_path))
+
+    k8s_cloud = await ops_test.add_k8s(kubeconfig=config, skip_storage=False)
+    k8s_model = await ops_test.track_model(
+        "cos", cloud_name=k8s_cloud, keep=ops_test.ModelKeep.NEVER
+    )
+    with ops_test.model_context("cos"):
+        yield k8s_model
```

Comment:
This should suffice.  No need to switch into the model_context.  It's kind of a convenience method to make things like `ops_test.model` and `ops_test.full_model_name` work for a selected model.  Think of `model_context` like juju switch.  when you `juju status` without a `-m`, you get the currently selected model.  But if you have the model_name, you can always `juju status -m $model`.
```suggestion
   yield k8s_model
```